scrap_emails.py

```diff
@@ -0,0 +1,125 @@
+
+
+    def parse(self, response):
+        """The main parsing logic of this crawler.
+
+        Args:
+            response: The response from resolving a remote url.
+
+        Returns:
+            A hash of {email: email_value} that is aggregated into an end output.
+        """
+
+        if self.verify_redirect_endpoint and not response.url.startswith(tuple(self.start_urls)):
+            return
+
+        for email in self.get_emails(response.text):
+            yield {'email': email}
+        for link in response.css('a::attr(href)').extract():
+            link = response.urljoin(link)
+
+            #yield {'link': link, 'start_url': self.start_urls[0]}
+
+            # First check as the setting doesn't seem to work as expected in scrapy. Will keep trying to load links after
+            # MAX_DEPTH even if it isn't actually allowing them to process further.
+            if response.meta["depth"] < self.custom_settings['DEPTH_LIMIT'] and self.is_valid_link(link) is True:
+
+                # Add this url to the visited list. Exclude any trailing slash marks for consistency.
+                self.existing_visited.append(re.sub(r'\/$', '', link))
+                yield scrapy.Request(link, callback=self.parse,)
+
+
+    def is_valid_link(self, l):
+        guessed_type = mimetypes.guess_type(l.split('?')[0])
+        # Ensure that we don't parse things like images or pdfs
+        if guessed_type[0] is None or guessed_type[0].startswith('text'):
+            # Ensure that we stick only to this domain.
+            # FIXME: There is an issue with http vs https in this sample application. Also with www vs non-www.
+            if l.startswith(tuple(self.start_urls)):
+                # Exclude sites we have already visited
+                if l not in self.existing_visited:
+                    return True
+
+        return False
+
+
+    def get_emails(self, s):
+        """Returns an iterator of matched emails found in string s."""
+        # Removing lines that start with '//' because the regular expression
+        # mistakenly matches patterns like 'http://foo@bar.com' as '//foo@bar.com'.
+        # Additionally, remove things that match a mimetype since that is generally an image link that looks like an email.
+        # Do some cleaning of the returned email. In this example case, only swap " at " for "@".
+        # Final check to not at some numbers that were matching the pattern.
+        return_emails = []
+        for email in re.findall(self.regex, s):
+            if email[0] not in self.matched_emails:
+                if not email[0].startswith('//'):
+                    if  mimetypes.guess_type(email[0])[0] is None or (not mimetypes.guess_type(email[0])[0].startswith('image') and not mimetypes.guess_type(email[0])[0].startswith('video')):
+                        cleaned_email = re.sub(r'\sat\s', '@', email[0])
+                        if not re.match(r'.+@[\d\.]+$', cleaned_email):
```

Comment:
I've split this up into two methods:

1. A method to clean up a matched email so that it is standardized in a "<email>@<domain>" format.

2. A method for determining if the regex matched email was a correctly matched email.

This reduces the nested logic in that one method. Additionally, it makes it easier to expand the code to include more variations one might try to write their email as on a website.

Not sure if there is more I should do to make it more readable?